tortoise/organizer/organizer.go

```diff
@@ -0,0 +1,143 @@
+package organizer
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/spacemeshos/go-spacemesh/common/types"
+	"github.com/spacemeshos/go-spacemesh/log"
+)
+
+// Callback is used to send iterated layers to the caller.
+// For siplicity Callback iterates over all layers unconditionally.
+type Callback func(types.LayerID)
+
+// Opt is for configuring Organizer.
+type Opt func(*Organizer)
+
+// WithLogger configures logger for Organizer.
+func WithLogger(logger log.Log) Opt {
+	return func(o *Organizer) {
+		o.logger = logger
+	}
+}
+
+// WithLastLayer layer configures last layer that tortoise knows about.
+func WithLastLayer(lid types.LayerID) Opt {
+	return func(o *Organizer) {
+		o.submitted = lid
+		o.received = lid
+	}
+}
+
+// WithWindowSize configures size of the window.
+// By default window of the size 1000 is used.
+func WithWindowSize(size int) Opt {
+	return func(o *Organizer) {
+		o.buf = make(layerBuffer, size)
+	}
+}
+
+// New creates Organizer instance.
+func New(opts ...Opt) *Organizer {
+	o := &Organizer{
+		logger: log.NewNop(),
+		buf:    make(layerBuffer, 1000), // ~4kb
+	}
+	for _, opt := range opts {
+		opt(o)
+	}
+	return o
+}
+
+// Organizer is a decorator for tortoise that keeps layers in order within the window
+// and sends them to tortoise only when the window has no gaps.
+type Organizer struct {
+	logger log.Log
+
+	// last submitted layer
+	submitted types.LayerID
+	// last received layer that is not submitted
+	received types.LayerID
+	// buf stores all layers between submitted and received
+	// buf is non-empty only if there are gaps after submitted
+	buf layerBuffer
+}
+
+// Iterate allows to iterate over all ordered layers that are currently within the window including new layer.
+func (o *Organizer) Iterate(ctx context.Context, lid types.LayerID, f Callback) {
+	if !lid.After(o.submitted) {
+		return
+	}
+	if lid.After(o.submitted.Add(uint32(len(o.buf)))) {
+		o.logger.WithContext(ctx).With().Error("tortoise organizer received a layer that does not fit in the current window."+
+			" subbmiting pending with gaps",
+			log.Int("window", len(o.buf)),
+			log.Named("submitted", o.submitted),
+			log.Named("received", lid),
+		)
+		o.submitAll(f)
+		// after all layers are submitted update submitted to the layer before this one
+		o.submitted = lid.Sub(1)
```

Comment:
> i don't understand why you have to overwrite this to lid.Sub(1). isn't it updated correctly in submitAll()?

submitAll updates only to the last layer that was received, and since there is no room for new layer o.received can't be updated beforehand

 > say o.submitted is 100 and lid is 1200, and o.received is 1150, don't we now have a gap (after submitAll()) btwn 1151 and 1200?

yes there will be an arbitrary number of gaps between 100 and 1200, including one between 1151 and 1200, hence there is an ERROR log here